horton/gbasis/cext.pyx

```diff
@@ -912,124 +907,219 @@ cdef class GOBasis(GBasis):
+        points : np.ndarray, shape=(npoint, 3), dtype=float
+            Cartesian grid points.
+        output : np.ndarray, shape=(npoint, 3), dtype=float
+            Output array. When not given, it is allocated and returned.
 
-           dm
-                A density matrix. For now, this must be a DenseTwoIndex object.
+        Returns
+        -------
+        output : np.ndarray, shape=(npoint, 3), dtype=float
+            The output array.
+        """
+        if output is None:
+            output = np.zeros((points.shape[0], 3), float)
+        self._compute_grid1_dm(dm, points, GB1DMGridGradientFn(self.max_shell_type), output)
+        return output
 
-           points
-                A Numpy array with grid points, shape (npoint,3).
+    def compute_grid_gga_dm(self, dm,
+                            np.ndarray[double, ndim=2] points not None,
+                            np.ndarray[double, ndim=2] output=None):
+        """Compute the electron density and gradient on a grid for a given density matrix.
 
-           **Optional arguments:**
+        **Warning:** the results are added to the output array! This may be useful to
+        combine results from different spin components.
 
-           output
-                A Numpy array for the output, shape (npoint,3). When not given,
-                it will be allocated.
+        Parameters
+        ----------
 
-           epsilon
-                Allow errors on the density of this magnitude for the sake of
-                efficiency.
+        dm : DenseTwoIndex
+            Density matrix, assumed to be symmetric.
+        points : np.ndarray, shape=(npoint, 3), dtype=float
+            Cartesian grid points.
+        output : np.ndarray, shape=(npoint, 4), dtype=float
+            Output array. When not given, it is allocated and returned. The first column
+            contains the density. The last three columns contain the gradient.
 
-           **Warning:** the results are added to the output array! This may
-           be useful to combine results from different spin components.
-        '''
+        Returns
+        -------
+        output : np.ndarray, shape=(npoint, 4), dtype=float
+            The output array.
+        """
         if output is None:
-            output = np.zeros((points.shape[0], 3), float)
-        self._compute_grid1_dm(dm, points, GB1DMGridGradientFn(self.max_shell_type), output, epsilon)
+            output = np.zeros((points.shape[0], 4), float)
+        self._compute_grid1_dm(dm, points, GB1DMGridGGAFn(self.max_shell_type), output)
         return output
 
     def compute_grid_kinetic_dm(self, dm,
                                 np.ndarray[double, ndim=2] points not None,
                                 np.ndarray[double, ndim=1] output=None):
-        '''Compute the kinetic energy density on a grid for a given density matrix.
+        """Compute the kinetic energy density on a grid for a given density matrix.
```

Comment:
It may seem strange, but it comes from the math:  the kinetic energy density is ambiguous but this specific form (and no other) shows up as the curvature of the exchange hole near electron coalescence, r=r'. The ability to get not only the depth and the (average) normalization of the exchange hole, but also its curvature (ergo, kind of, its width) correct is the whole motivation of meta-GGAs. The math of this is rather buried, old papers (ca. 1986) from Becke, Berkowitz, and Parr. But it all works out, and is a fun exercise for interested people.